arctic/store/version_store.py

```diff
@@ -576,6 +575,135 @@ def write(self, symbol, data, metadata=None, prune_previous_version=True, **kwar
         return VersionedItem(symbol=symbol, library=self._arctic_lib.get_name(), version=version['version'],
                              metadata=version.pop('metadata', None), data=None)
 
+    def _add_new_version_using_reference(self, symbol, new_version, reference_version, prune_previous_version):
+        if not new_version or \
+                not reference_version or \
+                new_version['symbol'] != reference_version['symbol'] or \
+                new_version['_id'] == reference_version['_id'] or \
+                not new_version['base_version_id']:
+            raise ArcticException("Illegal or non-matching nwe_version or reference_version were supplied: %s and %s",
+                                  new_version, reference_version)
+
+        # Check if in the meanwhile the reference version (based on which we updated incrementally) has been removed
+        last_look = self._versions.find_one({'_id': reference_version['_id']})
+        if last_look is None or last_look.get('deleted'):
+            raise pymongo.errors.OperationFailure("Failed to write metadata for symbol %s. "
+                                                  "The previous version (%s, %d) has been removed during the update" %
+                                                  (symbol, str(reference_version['_id']), reference_version['version']))
+
+        # There is always a small risk here another process in between these two calls (above/below)
```

Comment:
Couldn't you check that the `reference_version` exists after inserting the `new_version`?

Reply:
sure, I can check after and if missing I 'll remove the new version and raise an exception.

There will always be some extreme corner case scenarios (*) (since these actions are not atomic, or part of a transaction) however,  this check after insertion will certainly improve things.

process_B remains block just before the command which deletes the reference version, process_A finishes, process_B deletes the reference version

Reply:
Check is now pushed